### PR TITLE
Hexagonal lattice iterators

### LINKED ISSUE #2798: Segfault when trying to write tallies.out with DistribcellFilter
## Bug Description
Originally, I created a [forum post](https://openmc.discourse.group/t/standard-debugging-not-giving-more-info-than-runtimeerror-openmc-aborted-unexpectedly/3551/2) about a segmentation fault that occurred for my model. Included is the ordinary error and a more detailed stack trace using the debugger. That model is a bit complicated, so I created a simpler example that reproduces the error.

The segfault occurs when when I use a `DistribcellFilter` on a cell representing the inside of  an infinite `ZCylinder`. This cell appears in the model as part of a universe which is repeated in a `HexLattice` and is eventually intersected with `openmc.ZPlane`s to make a geometry with finite cells.

I'm pretty sure the error happens during the process of writing `tallies.out`, since the error vanishes when I tell my settings not to produce the `tallies.out` file. 

## Steps to Reproduce
Running this script will produce the error 
```
import openmc
import openmc.model
import numpy as np

radius = 0.9 # cm
pin_lattice_pitch = 2.0 # cm

# zplanes to define lower and upper region
z_low = openmc.ZPlane(z0=0,boundary_type='vacuum') # z=0 cm
z_mid = openmc.ZPlane(z0=10) # z=10 cm
z_high = openmc.ZPlane(z0=20,boundary_type='vacuum') # z=20 cm
hex_boundary_region = openmc.model.hexagonal_prism(edge_length=pin_lattice_pitch,orientation='y', boundary_type='periodic')

# materials
nat_U = openmc.Material(name="nat_U")
nat_U.set_density('g/cm3',1.0)
nat_U.add_element('U',1.0)

graphite = openmc.Material(name="Graphite")
graphite.set_density('g/cm3', 1.1995)
graphite.add_element('C', 1.0)

mats = openmc.Materials([nat_U,graphite])
mats.export_to_xml()

# cylinder
cyl = openmc.ZCylinder(r=radius)
cell = openmc.Cell(region=-cyl, fill=nat_U)
outer_cell = openmc.Cell(region=+cyl, fill=graphite)
univ = openmc.Universe(cells=[cell,outer_cell])

# make outer graphite universe to give to lattice
all_graphite_cell = openmc.Cell(fill=graphite)
outer_graphite_universe = openmc.Universe(cells=(all_graphite_cell,))

# create a hexagonal lattice of compacts
active_core_lattice = openmc.HexLattice()
active_core_lattice.pitch = (pin_lattice_pitch,)
active_core_lattice.center = (0., 0.)
active_core_lattice.outer = outer_graphite_universe

center = [univ]
ring = [univ, univ, univ, univ, univ, univ]

active_core_lattice.universes = [ring, center]
lower_hex_boundary_cell = openmc.Cell(fill=active_core_lattice, region=hex_boundary_region & +z_low & -z_mid)
upper_hex_boundary_cell = openmc.Cell(fill=active_core_lattice, region=hex_boundary_region & +z_mid & -z_high)
hex_cells = [lower_hex_boundary_cell,upper_hex_boundary_cell]

geometry = openmc.Geometry(hex_cells)
geometry.export_to_xml()

tallies = openmc.Tallies()

# moderator
tally = openmc.Tally(name="tally")
tally.scores = ['flux']
filter = openmc.DistribcellFilter(cell)
tally.filters = [filter]
tallies.append(tally)
tallies.export_to_xml()

# settings
settings = openmc.Settings()
# source definition. fission source given bounding box of graphite active region
system_LL = (-pin_lattice_pitch*np.sqrt(3)/2, -pin_lattice_pitch, 0)
system_UR = (pin_lattice_pitch*np.sqrt(3)/2, pin_lattice_pitch, 20)
source_dist = openmc.stats.Box(system_LL, system_UR, only_fissionable=True)
source = openmc.Source(space=source_dist)
settings.source = source
# k-eigenvalue settings
settings.particles = 10000
settings.inactive = 25
settings.batches = 100
settings.export_to_xml()

openmc.run()
```

## Environment
I am using OpenMC version 0.13.3 on Ubuntu 20.04.6 LTS. I installed from source in debug mode. The simulation is run using python 3.10.5.

### PR DESCRIPTION
**Pull request by pshriwise**
# Description

This contains a fix for the distribcell indexing of hexagonal lattices. It primarily comes into play when a hex lattice is used as a fill for more than one cell. In this case, we aren't correctly indexing into the lattice offset table, used to track the instance number of a universe (and its cells) in the lattice.

For some background, the lattice offsets for a hexagonal lattice contains some padding so that the 2-D indexing of the rings can be treated as a square array. This leaves some empty entries in the array. As a result, we need to jump forward a few entries to get to the next valid entry in the lattice offsets data member.

This is caused by the use of the same iterators for the `RectangularLattice` and `HexagonalLattice` classes. The `RectangularLattice` class uses the same iterator without accounting for this offset. This PR mainly adds an iterator that is specific to the `HexagonalLattice` class, which accounts for the empty entries based on the structure of the lattice rings.

I've added a test querying cell instances at the center of a hex lattice that fills multiple cells to ensure the correct instance values are calculated in an initialized simulation.

Fixes #2798 

# Checklist

- [ ] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
~- [ ] I have made corresponding changes to the documentation (if applicable)~
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by pshriwise** [comment]
@lewisgross1296

**Comment by gridley** [comment]
@pshriwise do you still consider this to be a draft?

File: src/geometry_aux.cpp [hunk 1]
```diff
@@ -530,7 +530,8 @@ std::string distribcell_path_inner(int32_t target_cell, int32_t map,
     if (c.type_ != Fill::MATERIAL) {
       int32_t temp_offset;
       if (c.type_ == Fill::UNIVERSE) {
-        temp_offset = offset + c.offset_[map];
+        temp_offset =
+          offset + c.offset_[map]; // should also apply to lattice fills?
```
Comment (pshriwise) [inline src/geometry_aux.cpp:534]:
```suggestion
          offset + c.offset_[map]; //TODO: should also apply to lattice fills?
```

**Comment by pshriwise** [comment]
> @pshriwise do you still consider this to be a draft?

I've been hesitant about this because the whole of our cell instance bookkeeping needs an overhaul IMO and I kinda feel like this is kicking the can down the road a bit. All the same, it does correct some behavior and we should probably get the fix in. Removing draft status now.

**Review by paulromano — APPROVED** [review]
Thanks for the fix @pshriwise! Sorry it took some time to get this one in

**Pull request merged**